WIP: exo clone: command to clone application and its service

martinjaime opened the pull request:
Addresses issue #27.

Opening pull request now to receive feedback while I write tests.
@kevgo, @rschifflin

ChrisMM commented on features/clone.feature:
```diff
@@ -0,0 +1,18 @@
+  an application onto my machine in one simple step so that I can work on it
+  without wasting time with setup activities.
+
+  - `exo clone [url]` clones an Exosphere application given a repository url.
```
Add the word "Rules" above these bullets.

kevgo replied:
Yes, that is how the official guidelines are. Over time, however, we came to the conclusion that "Rules:" can be optimized away without really losing readability, if its the only block of code. We'll add it back if there is also a "Notes:" section. So for the time being this is good to go from my point of view. Does that work for you @ChrisMM?

ChrisMM replied:
Wonderfully! 👍

kevgo replied:
@martinjaime the convention `[url]` means you can type "url" here or not, its up to you. What you want is `<url>`, which means you have to type an actual URL here, and its not optional. 

For completeness, making the URL optional would look like `[<url>]`. See http://docopt.org.

ChrisMM commented on features/clone.feature:
```diff
@@ -0,0 +1,18 @@
+Feature: cloning an Exosphere application and its services
+
+  As an application developer, I want to download the complete source code for
```
Formatting suggestion:

As an application developer
I want to download the complete source code for an application onto my machine in one simple step
So that I can work on it without wasting time with setup activities.

kevgo replied:
👍 there should be at least two newlines in there

ChrisMM commented on features/clone.feature:
```diff
@@ -0,0 +1,18 @@
+
+  Scenario: clone the "test" application with external dependencies.
+    Given I am in an empty folder
+    When starting "exo clone git@github.com:Originate/exosphere--example-app--space-tweet.git" in the terminal
```
When I run "exo clone git@github.com:Originate/exosphere--example-app--space-tweet.git" in the terminal

martinjaime replied:
The reason I worded it this was was because there are already steps in [features/steps/steps.ls](https://github.com/Originate/exosphere-sdk/blob/mj-issue27/features/steps/steps.ls) that use this wording. I was taking advantage of this so that no steps need to be added.

ChrisMM replied:
👍

kevgo replied:
The current logic behind the language is:
- avoid ego-centric steps. Instead of "When __I run__ xxx", say "when __running__ xxx". I think it reads better, more humble, and more like prose.
- we distinguish between __running__ a command (i.e. starting it and not continuing the script until it has finished) and __starting__ a command (i.e. kicking it off, then doing something with the running command in a later step, like entering text into it)

ChrisMM commented on features/clone.feature:
```diff
@@ -0,0 +1,18 @@
+  Scenario: clone the "test" application with external dependencies.
+    Given I am in an empty folder
+    When starting "exo clone git@github.com:Originate/exosphere--example-app--space-tweet.git" in the terminal
+    And waiting until I see "successfully cloned" in the terminal
```
Then I see "successfully cloned" in the terminal

ChrisMM commented on features/clone.feature:
```diff
@@ -0,0 +1,18 @@
+    Given I am in an empty folder
+    When starting "exo clone git@github.com:Originate/exosphere--example-app--space-tweet.git" in the terminal
+    And waiting until I see "successfully cloned" in the terminal
+    And waiting until I see "Done" in the terminal
```
And I see "Done" in the terminal

martinjaime replied:
The equivalent here, would be the step, `Then it prints "Done" in the terminal`. But this step only checks the printed output up until that point and does not wait for further output. There is more output printed before `Done` is finally printed.

kevgo replied:
The current test setup leaves the process running. I think the right course of action is to wait until the process has finished on its own, then verify the existing output of it.

martinjaime commented:
CircleCI is down. Tests for other features are not passing.

martinjaime commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,88 @@
+      @git-command process.cwd!, ['clone', @repo.url], (err) ~>
+        if err
+          @emit 'app-clone-failed', err
+          return @emit 'output', 'Done'
```
Maybe we shouldn't return and quit here. If the application already exists locally, it shouldn't quit since it doesn't necessarily mean that the services in application.yml already exist.

ChrisMM replied:
Let's be sure that case is covered in the feature, too.

kevgo replied:
Good point @martinjaime. Our script should behave a bit more intelligent. If an external service already exists, for example because another app also uses that service, it should not even try to clone the service again. It should just skip it, and tell the user that ("I didn't clone foo-service, because I found one already on the hard drive, and use that"), so that the user can verify its the correct service.

kevgo replied:
Please also remove emitting output here, and simplify similar to above.

kevgo commented on features/clone.feature:
```diff
@@ -0,0 +1,18 @@
+  without wasting time with setup activities.
+
+  - `exo clone [url]` clones an Exosphere application given a repository url.
+  - the URL must point to a Git repository containing an application.yml file at
```
`- the URL must point to a Git repository containing an Exosphere application, i.e. have an application.yml file in the root of the repository.`

kevgo commented on package.json:
```diff
@@ -9,8 +9,10 @@
+    "gitgo": "0.5.3",
     "inquirer": "1.1.2",
     "js-yaml": "3.6.1",
+    "lodash" : "4.13.1",
```
Please use http://www.preludels.com instead of lodash. Lodash is made for procedural coding, prelude is much better for functional-ish coding styles, and we already use it in this project. No point in loading two large utility functions that do the same thing  :)

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,88 @@
+  (@repo) ->
+
+
+  start-cloning: ->
```
Let's just call this method `start`, that's shorter and expresses the same thing. It makes sense that `AppCloner.start` starts cloning the app.

martinjaime replied:
@kevgo, that naming convention conforms with the classes of the other commands, should we deviate from that?

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,88 @@
+
+
+  start-cloning: ->
+    @verify-git-repo @repo.url, @verify-file-in-repo @repo.musthave-file, (err) ~>
```
For the sake of making this code self-describing, `verify-git-repo` sounds too generic. What exactly does this command verify? That it is a Git repo? Then it should be called `verify-is-git-repo`. That the repo already exists? Then it should be called `verify-git-repo-exists`. Etc.

martinjaime replied:
Will change to `verify-git-repo-exists`

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,88 @@
+    @verify-git-repo @repo.url, @verify-file-in-repo @repo.musthave-file, (err) ~>
+      if err
+        @emit 'app-verification-failed', err
+        return @emit 'output', 'Done'
```
No need to print `Done` if there was an error. The user knows the script ended, and won't do anything else anymore. So we don't have to tell them that again. To me, "done" means I'm done successfully. I got all the way to the end, and this is not a premature end of the program. We haven't reached that point here, so let's not print "done" in this case. 

Doing this allows us to simplify this to:

```ls
| err  =>  @emit 'app-verification-failed', err
```

The CLI class can then perform the output on the console explaining the error to the user.

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,88 @@
+        cloners = for service-name in Object.keys @app-config.services
+          service-dir = path.join @repo.path, @app-config.services[service-name].location
+          service-url = @app-config.services[service-name].url
+          new ServiceCloner service-name, {root: service-dir, url: service-url }
```
You can omit the curly braces here

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,88 @@
+          new ServiceCloner service-name, {root: service-dir, url: service-url }
+            ..on 'service-clone-success', (name) ~> @emit 'service-clone-success', name
+            ..on 'service-clone-fail', (name) ~> @emit 'service-clone-fail', name
+            ..on 'output', ({name, text}) ~> @emit 'output', text
```
Should we emit the service name here as well? It would be nice to know to which service the output belongs. An Exosphere app can have several dozen services!

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,88 @@
+            ..on 'output', ({name, text}) ~> @emit 'output', text
+        async.series [cloner.start for cloner in cloners when cloner.config.url], (err, exit-codes) ~>
+          | err                            => @emit 'all-clones-failed'
+          | @_contains-non-zero exit-codes => @emit 'some-clones-failed'
```
Please use two spaces around `=>`. It separates independent code constructs, and is more a control flow element.

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,88 @@
+            ..on 'service-clone-fail', (name) ~> @emit 'service-clone-fail', name
+            ..on 'output', ({name, text}) ~> @emit 'output', text
+        async.series [cloner.start for cloner in cloners when cloner.config.url], (err, exit-codes) ~>
+          | err                            => @emit 'all-clones-failed'
```
What do you think about calling this `clone-all-failed`? Its that that all clone operations have failed, just that the "clone all" operation has failed.

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,88 @@
+          | err                            => @emit 'all-clones-failed'
+          | @_contains-non-zero exit-codes => @emit 'some-clones-failed'
+          | otherwise                      => @emit 'all-clones-successful'
+          @emit 'output', 'Done'
```
Let's keep encapsulation. This class (as well as all other "inner" classes that perform business logic) shouldn't be concerned with user interaction at all. User interaction includes outputting text. Only the "CLI" class should interact with the shell and the user. 

Emitting an event that causes the CLI class to print something is similar to printing it yourself here. The inner classes should just emit events that signify lifecycle events ("I'm starting to clone", "I'm done cloning", etc).

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,88 @@
+    gitgo(path, command)
+      ..on 'error', (error) ->
+        done error
+        ..unpipe!
```
This reads strange. What is `unpipe` called on? Better to make this a bit more explicit using a temp variable here.

```ls
git-process = gitgo(path, command)
  ..on 'error', (error) ->
    git-process.unpipe!
    done error

kevgo replied:
Why do you have to call `unpipe` here?

martinjaime replied:
It was to avoid the process from continuing printing to the console. but a git command usually stops printing once it errors, so I will remove this line.

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,88 @@
+      ..on 'error', (error) ->
+        done error
+        ..unpipe!
+      ..on 'finish', -> done null
```
`done!` is the same as `done null`, but shorter

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,88 @@
+        done error
+        ..unpipe!
+      ..on 'finish', -> done null
+      ..pipe process.stdout
```
In the next PR, please pipe this through https://github.com/Originate/exosphere-sdk/blob/master/src/logger.ls, so that the output is properly colored and prefixed with the service name.

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,88 @@
+
+
+  verify-git-repo: (url, callback) ->
+    path = '/tmp/gitgo_temp_repo-' + new Date()
```
Please use https://github.com/raszi/node-tmp to get a temp directory.

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,88 @@
+    path = '/tmp/gitgo_temp_repo-' + new Date()
+    fs.mkdirSync(path)
+    async.series [
+      (next) ~> @git-command path, ['init'], next
```
Why do you initialize Git repos in folders here? It shouldn't do that. If `exo clone` finds a folder that is not a Git repo, it should abort and tell the user that there is a folder with a conflicting name there which is most likely not the service that the application expects.

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,88 @@
+    (err, path) ~>
+      return done err if err
+      args = ['cat-file', '-e', 'origin/master:' + file-name]
+      @git-command path, args, done
```
Why can't you simply 
- ensure that the application repo doesn't exist yet (abort if it does - no need to clone it)
- clone the application repo onto the users machine
- check the cloned application folder and make sure it has the right files
- if not, abort and delete the app folder
- check for existing services and make sure they actually are Exosphere services (are a Git repo and contain `service.yml`)
- clone the missing services

martinjaime commented on src/logger.ls:
```diff
@@ -24,6 +23,12 @@ class Logger
+  set-colors: (service-names) ->
+    for service-name, i in service-names
+      @colors[service-name] = Logger._default_colors[i % Logger._default_colors.length]
+    @length = map (.length), Object.keys(@colors) |> maximum
```
Created this `set-colors` method so that `@colors` can still be initiated after the constructor of Logger is called. This way, Logger does not depend on a preexisting application.yml file. It does not change its previous functionality, and added this feature.

kevgo replied:
I like it!

kevgo replied:
Totally nitpicking and geeking out on style here. What do you think about this version?

```ls
@length = Object.keys(@colors).map (.length) |> maximum

martinjaime commented on src/logger.ls:
```diff
@@ -24,6 +23,12 @@ class Logger
+    @length = map (.length), Object.keys(@colors) |> maximum
+
+
   @_default_colors = [magenta, cyan, yellow]
```
Should we add a fourth color to this list, since exo-clone was added? This is why I am modding the index by the size of this list on line 28 of this script.

martinjaime replied:
@kevgo, could you please comment on the changes in this file?

kevgo replied:
Yes, please add more colors here.

kevgo commented on src/commands/clone/cli.ls:
```diff
@@ -0,0 +1,36 @@
+console.log 'We are going to clone an Exosphere application!\n'
+
+[_, _, _, repo-url] = process.argv
+return missing-url! unless repo-url
```
Where is this method defined?

trushton replied:
@kevgo [here](https://github.com/Originate/exosphere-sdk/blob/bb252e74ce802bc8f447f4a75059e477077a947c/src/commands/clone/cli.ls#L45)

kevgo replied:
That appears to be the same file as this one. Why isn't this function in here then?

kevgo commented on src/commands/clone/cli.ls:
```diff
@@ -0,0 +1,36 @@
+[_, _, _, repo-url] = process.argv
+return missing-url! unless repo-url
+
+[..._, repo-name] = repo-url.split('/')
```
Can you remove the `_` here? (I'm not sure)

kevgo replied:
Remove parentheses

kevgo replied:
And instead of manual juggling of filesystem strings, can you use https://nodejs.org/api/path.html#path_path_basename_path_ext here?

martinjaime replied:
Changed to `path.basename 'repo-name'`

kevgo commented on src/commands/clone/cli.ls:
```diff
@@ -0,0 +1,36 @@
+return missing-url! unless repo-url
+
+[..._, repo-name] = repo-url.split('/')
+repo-name = repo-name.replace '.git', ''
```
Use accessignment here!

```ls
repo-name .= replace '.git', ''

kevgo commented on src/commands/clone/cli.ls:
```diff
@@ -0,0 +1,36 @@
+
+[..._, repo-name] = repo-url.split('/')
+repo-name = repo-name.replace '.git', ''
+repo-path = process.cwd! + '/' + repo-name
```
Use `path.join` to concatenate path names together in platform-independent ways (`/` doesn't work on Windows)

kevgo commented on src/commands/clone/cli.ls:
```diff
@@ -0,0 +1,36 @@
+
+logger = new Logger
+
+app-cloner = new AppCloner repo
```
I think you can inline the `repo` variable here

martinjaime replied:
I've made changes in commit 13ec9f6895d23fb487293920d4677344ea046704 where it needs to reference app-cloner in the listeners.

kevgo commented on src/commands/clone/cli.ls:
```diff
@@ -0,0 +1,36 @@
+logger = new Logger
+
+app-cloner = new AppCloner repo
+  ..on 'output', (text) -> logger.log name: 'exo-clone', text: text
```
This line goes away

kevgo commented on src/commands/clone/cli.ls:
```diff
@@ -0,0 +1,36 @@
+app-cloner = new AppCloner repo
+  ..on 'output', (text) -> logger.log name: 'exo-clone', text: text
+  ..on 'app-verification-failed', (err) -> logger.log name: 'exo-clone', text: red "Error: repo could not be verified\n#{err}"
+  ..on 'app-clone-success' -> logger.log name: 'exo-clone', text: "Application #{@repo.name} successfully cloned."
```
Remove "successfully" and the period (this is not a proper sentence

kevgo commented on src/commands/clone/cli.ls:
```diff
@@ -0,0 +1,36 @@
+  ..on 'app-clone-success' -> logger.log name: 'exo-clone', text: "Application #{@repo.name} successfully cloned."
+  ..on 'app-clone-failed', (err) -> logger.log name: 'exo-clone', text: red "Error: cloning the project failed\n#{err}"
+  ..on 'service-clone-fail', (name) -> logger.log name: 'exo-clone', text: red "Cloning #{name} failed"
+  ..on 'all-clones-failed', -> logger.log name: 'exo-clone', text: red "All Services failed to clone."
```
There is `all-clone-failed` and `all-clones-failed`. Can you come up with strings that are easier to distinguish?

martinjaime replied:
@kevgo, did you misread `app-clone-failed` and `all-clones-failed`?

kevgo commented on src/commands/clone/cli.ls:
```diff
@@ -0,0 +1,36 @@
+  ..on 'app-clone-failed', (err) -> logger.log name: 'exo-clone', text: red "Error: cloning the project failed\n#{err}"
+  ..on 'service-clone-fail', (name) -> logger.log name: 'exo-clone', text: red "Cloning #{name} failed"
+  ..on 'all-clones-failed', -> logger.log name: 'exo-clone', text: red "All Services failed to clone."
+  ..on 'some-clones-failed', -> logger.log name: 'exo-clone', text: yellow "Some Services failed to clone."
```
As a user, I don't see the difference between "some clones failed" and "all clones failed", and how that is a relevant distinction for me that leads to different courses of action. Can you either explain a bit more here, or just tell the user which services had issues with cloning, and what the issues were?

martinjaime replied:
@kevgo, if during the process of cloning an application and its services, some services fail to clone, `exo clone` will print regular git error messages that are descriptive. 'exo clone` will continue to try to clone the rest of the services even if some fail. This is why I chose to say "some" or "all", since the user will be able to see which services failed.

kevgo commented on src/commands/clone/cli.ls:
```diff
@@ -0,0 +1,36 @@
+  ..on 'service-clone-fail', (name) -> logger.log name: 'exo-clone', text: red "Cloning #{name} failed"
+  ..on 'all-clones-failed', -> logger.log name: 'exo-clone', text: red "All Services failed to clone."
+  ..on 'some-clones-failed', -> logger.log name: 'exo-clone', text: yellow "Some Services failed to clone."
+  ..on 'all-clones-successful', -> logger.log name: 'exo-clone', text: green "Services successfully cloned."
```
Here is where it should print "done"  😄

martinjaime replied:
@kevgo, see [this commit](https://github.com/Originate/exosphere-sdk/pull/83/commits/13ec9f6895d23fb487293920d4677344ea046704) for update.

kevgo commented on src/logger.ls:
```diff
@@ -1,29 +1,34 @@
 
+  service-color: (name) ->
+    @colors[name] ?= Logger._default_colors[_.size(@colors) % Logger._default_colors.length]
+    @colors[name]
```
I think you can just remove this line and it should still work. The result of the previous line should be either `@colors[name]` or what you assign to it.

kevgo commented:
This is a great start! Let's see how it looks after all comments are addressed.

kevgo commented on features/clone.feature:
```diff
@@ -0,0 +1,19 @@
+  in one simple step so that I can work on it without wasting time with setup
+  activities.
+
+  - `exo clone <url>` clones an Exosphere application given a repository url.
```
Gherkin (the language used in `.feature` files) doesn't understand backticks. Please use double-quotes here.

kevgo replied:
Please remove the periods, these aren't proper sentences

kevgo replied:
"exo clone <url>" downloads the Exosphere application at the given URL onto the user's machine

kevgo commented on features/clone.feature:
```diff
@@ -0,0 +1,19 @@
+
+  Scenario: clone the "test" application with external dependencies.
+    Given I am in an empty folder
+    When starting "exo clone git@github.com:martinjaime/exo-sample-app.git" in the terminal
```
When __running__ "...
Then it prints "..." in the terminal

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,73 @@
+
+class AppCloner extends EventEmitter
+
+  (@repo) ->
```
Please use full words for variables, and make this one more descriptive. Is this the repo name? Then it should be called `repository-name`. Is it the repo URL? Then call it `repo-url` etc.

martinjaime replied:
`@repo` is an object with basic metadata about the application's repository. that is why it has a general name. I've changed it to a full word as `@repository`

kevgo replied:
If it contains data about the repository, what about naming it `@repository-data` or `@repo-data`?

kevgo commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,73 @@
+
+
+  start: ->
+    @verify-is-exo-repo @repo.url, @verify-file-in-repo @repo.musthave-file, (err) ~>
```
I think the steps I outlined in https://github.com/Originate/exosphere-sdk/pull/83/files#r71626070 make more sense here. If you disagree, let's hop on a call and talk about it.  :)

martinjaime replied:
I agree with you. I'm currently working on changing the process to your suggestion. I will push it now so you can see, but know it is not finalized. In this commit, I also removed the gitgo dependency. The only problem I am having now is having observable process actually send console output to the method I give it in service-cloner.ls.

kevgo replied:
If you provide the logger to ObservableProcess (via the `env` parameter), it will pipe output straight to the logger for you. You might have to inject the logger as a dependency to be able to do that. Please tag me when the changes are fully done. Thanks for all the great work here!

kevgo commented on src/logger.ls:
```diff
@@ -24,6 +23,12 @@ class Logger
       console.log color(bold "#{@_pad name} "), color(line)
 
 
+  set-colors: (service-names) ->
```
Please add a comment which describes that this method can be called again after initialization if more services are added.

kevgo commented:
The changes made so far look good. Please address some of the comments in app-cloner.ls

martinjaime commented:
@kevgo, latest [commit](https://github.com/Originate/exosphere-sdk/pull/83/commits/bb252e74ce802bc8f447f4a75059e477077a947c) addresses most issues, but still a work in progress.

kevgo commented:
No worries. Please tag me again when all comments have been addressed.  :)

martinjaime commented on src/commands/clone/app-cloner.ls:
```diff
@@ -0,0 +1,80 @@
+            ..on 'output', (data) ~> @emit 'output', data
+        async.series [cloner.start for cloner in cloners when cloner.config.url], (err, exit-codes) ~>
+          | err                                =>  @emit 'all-clones-failed'
+          | count = @_count-errors exit-codes  =>  @emit 'some-clones-failed', count, exit-codes.length
```
The output message from this event will display how many services failed to clone.

kevgo commented on src/logger.ls:
```diff
@@ -6,25 +6,33 @@ require! {
 
-  @_default_colors = [magenta, cyan, yellow]
+  set-colors: (service-names) ->
+  # This method may be called after initialization to set/reset colors, given a
```
Move the comment above the method name.

kevgo commented:
🚢
